R/utils_conditions.R

```diff
@@ -0,0 +1,735 @@
+#'   function, it will give its package's name, and if it is called from a
+#'   global function, will give the default `global_hit_result`
+#' @param global_hit_result optional; non-function or a `function(input_env,
+#'   current_env) <result>`; what result to return when there appears to be no
+#'   associated package, and the global environment is hit during the ancestor
+#'   search. This is the expected result when this function is called from code
+#'   outside of a package (but it may also happen if the caller grabbed `env`
+#'   from the call stack, and jumped the wrong number of stack frames). The
+#'   default is to raise an error in this case, but pass in a string here and it
+#'   will be returned instead.
+#' @param empty_hit_result optional; non-function or a `function(input_env,
+#'   current_env) <result>`; what result to return when there appears to be no
+#'   associated package, and the empty environment is hit during the ancestor
+#'   search. This is expected to happen if called on an environment default is
+#'   to produce an error.
+#' @param imports_hit_result optional; non-function or a `function(input_env,
+#'   current_env) <result>`; what result to return when there appears to be no
+#'   associated package, and a package imports environment is hit during the
+#'   ancestor search. This case is expected to occur only in advanced use when
+#'   there's been some issue with requesting the wrong ancestor of an
+#'   environment when preparing the `env` argument. Default is to produce an
+#'   error.
+#' @param exports_hit_result optional; non-function or a `function(input_env,
+#'   current_env) <result>`; what result to return when there appears to be no
+#'   associated package, and a package exports environment is hit during the
+#'   ancestor search. This is not expected to occur in any foreseen use case,
+#'   and might only be possible if, after the package is attached, `env` is
+#'   manually constructed as a descendant of the package exports environment.
+#'   Default is to produce an error.
+#' @param namespace_hit_result optional; non-function or a `function(input_env,
+#'   current_env) <result>`; what result to return when we successfully find a
+#'   package namespace environment among `env` or its ancestors;
+#'
+#' @return a string containing the associated package name, or one of the
+#'   overrides specified in {`global_hit_result`, `empty_hit_result`,
+#'   `imports_hit_result`, `exports_hit_result`, `namespace_hit_result`}
+#'
+#' @details The tests to detect a package imports and exports environments are
+#'   mostly name-based and thus may be a little imprecise, but also may be more
+#'   robust to any package-related environment restructuring that may happen (or
+#'   may have happened already) in other versions of R.
+#'
+#' @importFrom rlang caller_env is_reference global_env base_env empty_env
+#'   is_function
+#' @importFrom tibble lst
```

Comment:
Might need `@noRd` here and in other internal helper functions.  Not the handler ones though, as they will be referenced by the tutorial.

Reply:
vs. `@keywords internal`?